### ISSUE TITLE
Job endpoint

### ISSUE THREAD

**Issue by aldbr** [aldbr/dirac-cwl-proto#1]
Implement the job interface presented in https://github.com/DIRACGrid/diracx/discussions/175. 
Composed of a `typer CLI` that corresponds to a dirac client, another function would moke the dirac router.

- [ ] local submission
  - [x] input sandbox
  - [x] input data
  - [ ] output sandbox
  - [ ] output data 
  - [ ] make sure a job with local requirements (per step) cannot be processed (or at least a warning should be returned)
- [ ] dirac submission (translate to JDL before submitting) :arrow_right: would be integrated in `diracx` if it works well

**Comment by aldbr** [comment]
Moving job/transformation "Type" from `metadata` to `description`? I would say yes since we said we will handle metadata in a generic way.
Moving `metadata` section to `description`? Why not?

**Comment by natthan-pigoux** [comment]
> "Type" from `metadata` to `description`

Do you mean merging: https://github.com/aldbr/dirac-cwl-proto/blob/main/src/dirac_cwl_proto/submission_models.py#L80 ?

**Comment by aldbr** [comment]
Yes, we would have the `cwl workflow`, the `cwl parameters` and the `dirac job description` that would contain the `dirac metadata` (so 3 parameters instead of 4).

**Comment by natthan-pigoux** [comment]
Regarding output sandbox, do you think we add a specific job parameter to specify it?
Do we split the `sandbox` parameter into `input_sandbox` and `output_sandbox`?

**Comment by aldbr** [comment]
> Regarding output sandbox, do you think we add a specific job parameter to specify it?

We would need to a specific job parameter yes, something like: `output_sandbox: list[str]` with values such as `["output.sim"]`.

> Do we split the sandbox parameter into input_sandbox and output_sandbox?
> The sandbox parameter is currently a list, it could become: sandbox: dict[SandboxType, list] as: sandbox = {'input': ['input_*'], 'output': ['stdout', 'stderr']}

Very good question! Currently the parameters allow to create multiple jobs in a single request. The current `json` passed to the server:

```
- cwl task
- params1
- params2
- params3
```

Would generate 3 jobs such as:

```
- cwl task
- params1

- cwl task
- params2

- cwl task
- params3
```

In its current form, you can specify one input sandbox per job.
The question is: can we have different values inside the output sandboxes of the jobs that depend on the inputs?

For instance:
- in `params1`, `output_sandbox` would be `["output1.sim"]`
- in `params2`, `output_sandbox` would be `["output2.sim"]`
- in `params3`, `output_sandbox` would be `["something-completely-different"]`

If the answer is yes: then we should likely add `output_sandbox` within the cwl parameters.
Else: then we should add `output_sandbox` in the job description model.

Note: `params1`, and `params2` could be treated as `output_sandbox=<output-name-from-cwl>`.

Is it clear? 😅 

Here is my point of view:
- I think that in general, we would always want to get the same values in the output sandboxes of all the jobs submitted within the same request
- It would add flexibility, and we could imagine a system where the default input/output sandboxes are given through the job description and can be overridden in the parameters, but I don't know how this could be nicely implemented.
  The current approach described in the [transition plan](https://github.com/DIRACGrid/diracx/discussions/175#discussioncomment-9595943) allows users to submit jobs by just providing their cwl workflows and inputs (and dirac-specific parameters, either as a separate yaml or directly within the cwl).
`diracx-cli` then finds the files available locally and create input sandboxes to add them to the request before sending it to `diracx-routers`.

**Comment by natthan-pigoux** [comment]
> Is it clear? 😅
>

Yes it is ! thanks 😄 

> Here is my point of view:
> 
> * I think that in general, we would always want to get the same values in the output sandboxes of all the jobs submitted within the same request

Clearly, that would be the most common case.

> * It would add flexibility, and we could imagine a system where the default input/output sandboxes are given through the job description and can be overridden in the parameters, but I don't know how this could be nicely implemented.

I can try to think about it, and propose something.

>   The current approach described in the [transition plan](https://github.com/DIRACGrid/diracx/discussions/175#discussioncomment-9595943) allows users to submit jobs by just providing their cwl workflows and inputs (and dirac-specific parameters, either as a separate yaml or directly within the cwl).
>   `diracx-cli` then finds the files available locally and create input sandboxes to add them to the request before sending it to `diracx-routers`.

That is currently managed by the metadata models and the input yaml files (e.g. `type_dependencies`), correct?

**Comment by arrabito** [comment]
Yes I totally agree. My question was because in the current code dirac metadata are currently mixed.

**Comment by Loxeris** [comment]
Additionally we need a way to differentiate the outputs going into sandboxes and the ones going in the file catalog.
The list of outputs for the file catalog could be defined in parameter files and/or CWL hints. Does anyone have any other idea?

**Comment by mexanick** [comment]
Shouldn't we just abstract it in the DataCatalogInterface? Right now hardcoded "filecatalog" is a remnant of the old implementation, which now can be overloaded depending on the kind of the DataCatalogInterface. 

and considering what goes where, for this we have pre/post processing in the ExecutionHooks

**Comment by aldbr** [comment]
Closing that as we have played enough with the POC to start the ADR and focus on the implementation within diracx
